source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.cc

```diff
@@ -0,0 +1,96 @@
+#include "source/extensions/http/cache/file_system_http_cache/cache_file_header_proto_util.h"
+
+#include "absl/container/flat_hash_set.h"
+#include "absl/strings/str_cat.h"
+
+namespace Envoy {
+namespace Extensions {
+namespace HttpFilters {
+namespace Cache {
+namespace FileSystemHttpCache {
+
+CacheFileHeader mergeProtoWithHeadersAndMetadata(const CacheFileHeader& entry_headers,
+                                                 const Http::ResponseHeaderMap& response_headers,
+                                                 const ResponseMetadata& response_metadata) {
+  Http::ResponseHeaderMapPtr merge_headers = headersFromHeaderProto(entry_headers);
+  applyHeaderUpdate(response_headers, *merge_headers);
+  return makeCacheFileHeaderProto(entry_headers.key(), *merge_headers, response_metadata);
+}
+
+CacheFileHeader makeCacheFileHeaderProto(const Key& key,
+                                         const Http::ResponseHeaderMap& response_headers,
+                                         const ResponseMetadata& metadata) {
+  CacheFileHeader file_header;
+  *file_header.mutable_key() = key;
+  TimestampUtil::systemClockToTimestamp(metadata.response_time_,
+                                        *file_header.mutable_metadata_response_time());
+  response_headers.iterate([&file_header](const Http::HeaderEntry& header) {
```

Comment:
nit: you could factor out this iteration using a template function so you don't have to repeat it for trailers.

Reply:
It sounds like you were thinking of a larger factoring out than I've done, but I think what I've done here is the most factoring out that doesn't involve a really weird template function (because add_headers vs add_trailers, and either way it needs to be a lambda anyway for the captured destination)